specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+{
+    auto channelUri = result.Channel().Uri();
+    auto channelExpiry = result.Channel().ExpirationTime();
+
+    // Persist the channelUri and Expiry in the App Service
+}
+else if (result.Status() == ChannelStatus::CompletedFailure)
+{
+    LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
+}
+```
+
+## Manifest Registration
+For MSIX, the COM activator GUID and the exe path need to be registered in the manifest. The launch args would need to be pre set to a well-known string that defines Push Triggers. 
+```xml
+<Extensions>
+  <com:Extension Category="windows.comServer">
+    <com:ComServer>
+      <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="-PushServer">
+        <com:Class Id="{GUIDEntryPoint}" />
+      </com:ExeServer>
+    </com:ComServer>
+  </com:Extension>
+</Extensions>
+```
+
+## UWP InProc activator
+The UWP inproc activator flow differs from the Win32 inproc activator flow since there is technically no WinMain activation associated with UWPs. Instead the App will have to  override and implement OnBackgroundActivated which is tightly coupled to Windows.UI.Xaml workstreams.
+
+# API Details
+```c# (but really MIDL3)
+
+namespace Microsoft.ProjectReunion.PushNotifications
+{
+    // Event args in WinMain activation payload.
+    runtimeclass PushActivatedEventArgs
+    {
+        // Initialize using the raw byte payload
+        PushActivatedEventArgs(Windows.ApplicationModel.Background.IBackgroundTaskInstance backgroundTask);
+
+        // The Push payload
+        byte[] Payload { get; };
+
+        // Takes a deferral to run under specific modes like low power mode
+        void GetDeferral();
+
+        // Closes the Deferral taken on a Push Trigger. If this is not called by the developer, the deferral will automatically be invoked on destruction.
+        void CompleteDeferral();
+
+        // Subscribe to Cancelled event handler to be signalled when resource policies are no longer true like 30s wallclock timer
+        event Windows.ApplicationModel.Background.BackgroundTaskCanceledEventHandler Canceled;
+
+        // The number of times Resource management suspended the Task
+        UInt32 SuspendedCount { get; };
+    };
+
+    // An abstraction over the inproc activation Registration flow
+    runtimeclass InProcActivatorDetails
+    {
+        // Initialize using a manifest defined activatorId to Activate the InProc COM server
+        InProcActivatorDetails(Guid taskClsid);
+
+        // Initialize an inproc activator for LOW-IL apps like UWP with no COM Server specified in manifest
+        InProcActivatorDetails();
+
+        // The CLSID associated with the Client COM server that reunion will adopt
+        Guid TaskClsid{ get; };
+
+        // The conditions under which Push Triggers would execute
+        Windows.ApplicationModel.Background.IBackgroundCondition Condition;
```

Comment:
You can currently have [multiple conditions](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/set-conditions-for-running-a-background-task#place-multiple-conditions-on-your-background-task), like both UserPresent and InternetAvailable. So this would have to be an array.

Reply:
Ah. Great catch!  I will fix this.